runtime/vm/BytecodeInterpreter.hpp

```diff
@@ -5253,15 +5258,31 @@ class INTERPRETER_CLASS
 		_currentThread->callOutCount += 1;
 #endif /* JAVA_SPEC_VERSION >= 19 */
 		updateVMStruct(REGISTER_ARGS);
+#if JAVA_SPEC_VERSION >= 21
+		/* Remove VMaccess in the trivial downcall to benefit the JIT for better performance. */
+		if (!_currentThread->isInTrivialDownCall) {
+			VM_VMAccess::inlineExitVMToJNI(_currentThread);
+		}
+#else /* JAVA_SPEC_VERSION >= 21 */
 		VM_VMAccess::inlineExitVMToJNI(_currentThread);
+#endif /* JAVA_SPEC_VERSION >= 21 */
 		VM_VMHelpers::beforeJNICall(_currentThread);
 #if FFI_NATIVE_RAW_API
 		ffiCallWithSetJmpForUpcall(_currentThread, cif, function, returnStorage, values, values_raw);
 #else /* FFI_NATIVE_RAW_API */
 		ffiCallWithSetJmpForUpcall(_currentThread, cif, function, returnStorage, values);
 #endif /* FFI_NATIVE_RAW_API */
 		VM_VMHelpers::afterJNICall(_currentThread);
+#if JAVA_SPEC_VERSION >= 21
+		/* Remove VMaccess in the trivial downcall to benefit the JIT for better performance. */
+		if (!_currentThread->isInTrivialDownCall) {
+			VM_VMAccess::inlineEnterVMFromJNI(_currentThread);
+		}
+		/* Reset the value in any case after performing the downcall. */
+		_currentThread->isInTrivialDownCall = FALSE;
```

Comment:
This isn't sufficient; there are error paths that avoid this; suggest it belongs after the `done` label:
```c
done:
#if JAVA_SPEC_VERSION >= 21
		/* Clear the trivial downcall flag. */
		_currentThread->isInTrivialDownCall = FALSE;
#endif /* JAVA_SPEC_VERSION >= 21 */
```

Reply:
The VM exits/throw out exceptions if any error occurs after but I agreed to move after `done` for safety (just updated against the suggestion above).